core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt

```diff
@@ -21,43 +21,75 @@ private enum class JoinType(val type: String) {
-                "ON ${onConditionDopeQuery.queryString}",
-            parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + onConditionDopeQuery.parameters,
-        )
+        this.joinType = joinType
+        this.bucket = bucket
+        this.onCondition = onCondition
+        this.parentClause = parentClause
     }
 
     constructor(joinType: JoinType, bucket: Bucket, onKeys: Field<out ValidType>, parentClause: ISelectFromClause) {
-        val parentDopeQuery = parentClause.toDopeQuery()
-        val bucketDopeQuery = bucket.toDopeQuery()
-        val keyDopeQuery = onKeys.toDopeQuery()
-        dopeQuery = DopeQuery(
-            queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
-                "ON KEYS ${keyDopeQuery.queryString}",
-            parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + keyDopeQuery.parameters,
-        )
+        this.joinType = joinType
+        this.bucket = bucket
+        this.onKeys = onKeys
+        this.parentClause = parentClause
     }
 
     constructor(joinType: JoinType, bucket: Bucket, onKey: Field<out ValidType>, forBucket: Bucket, parentClause: ISelectFromClause) {
-        val parentDopeQuery = parentClause.toDopeQuery()
-        val bucketDopeQuery = bucket.toDopeQuery()
-        val keyDopeQuery = onKey.toDopeQuery()
-        val forBucketDopeQuery = forBucket.toDopeQuery()
-        dopeQuery = DopeQuery(
-            queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
-                "ON KEY ${keyDopeQuery.queryString} FOR ${forBucketDopeQuery.queryString}",
-            parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + keyDopeQuery.parameters + forBucketDopeQuery.parameters,
-        )
+        this.joinType = joinType
+        this.bucket = bucket
+        this.onKey = onKey
+        this.forBucket = forBucket
+        this.parentClause = parentClause
     }
 
-    override fun toDopeQuery() = dopeQuery
+    override fun toDopeQuery(): DopeQuery =
+        when {
+            onCondition != null -> {
+                val parentDopeQuery = parentClause.toDopeQuery()
+                val bucketDopeQuery = bucket.toDopeQuery()
+                val onConditionDopeQuery = onCondition!!.toDopeQuery()
+                DopeQuery(
+                    queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
+                        "ON ${onConditionDopeQuery.queryString}",
+                    parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + onConditionDopeQuery.parameters,
+                )
+            }
+
+            onKeys != null -> {
+                val parentDopeQuery = parentClause.toDopeQuery()
+                val bucketDopeQuery = bucket.toDopeQuery()
+                val keyDopeQuery = onKeys!!.toDopeQuery()
```

Comment:
please never use !!

Reply:
I used it, becuase i already check if the onKeys is not Null in the `when`, so i know that this can't throw an exception. 
When you do null checks in an `if else` then the IDEA notices that, but with the `when` it doesn't work

Reply:
done